### PR TITLE
Use rayon to parallelize the stability check

### PR DESCRIPTION
**Pull request by MichaReiser**
## Summary

This PR uses rayon to parallelize the stability check by scheduling each project as its own task.

## Test Plan

I ran the ecosystem check. It now makes use of all cores (except at the end, there are some large projects). 

## Performance

The check now completes in minutes where it took about 30 minutes before.

### CODE REVIEW

File: crates/ruff_dev/src/check_formatter_stability.rs [hunk 1]
```diff
@@ -84,8 +76,88 @@ pub(crate) fn main(args: &Args) -> anyhow::Result<ExitCode> {
...
+    let mut total_files = 0;
+    let start = Instant::now();
+
+    rayon::scope(|scope| {
+        let (sender, receiver) = channel();
+
+        for base_dir in &args.files {
+            for dir in base_dir.read_dir().unwrap() {
+                let path = dir.unwrap().path().clone();
+                sender.send(Message::Start { path: path.clone() }).unwrap();
```
Comment (konstin) [inline crates/ruff_dev/src/check_formatter_stability.rs:105]: i'd move that inside the spawn

File: crates/ruff_dev/src/check_formatter_stability.rs [hunk 2]
```diff
@@ -84,8 +76,88 @@ pub(crate) fn main(args: &Args) -> anyhow::Result<ExitCode> {
...
+                    }
+                    .unwrap();
+                });
+            }
+        }
+
+        scope.spawn(|_| {
+            let mut stdout = stdout().lock();
+
+            for message in receiver {
```
Comment (konstin) [inline crates/ruff_dev/src/check_formatter_stability.rs:125]: thread pools and generators would be a dream combination for writing this kind of code, basically you would write a `par_iter` and flatten over the messages

**Review by konstin — APPROVED** [review]

File: crates/ruff_dev/src/check_formatter_stability.rs (hunk 1 above)
Comment (MichaReiser) [reply crates/ruff_dev/src/check_formatter_stability.rs:105]: Much better. Thanks

**Comment by MichaReiser** [comment]
[Graphite](https://app.graphite.dev/github/pr/astral-sh/ruff/5441) rebased this pull request as part of a merge.

**Comment by MichaReiser** [comment]
@MichaReiser merged this pull request with [Graphite](https://app.graphite.dev/github/pr/astral-sh/ruff/5441).

**Pull request merged**